crates/noirc_frontend/src/hir_def/types.rs

```diff
@@ -51,6 +51,11 @@ pub enum Type {
     /// represents the generic arguments (if any) to this struct type.
     Struct(Shared<StructType>, Vec<Type>),
 
+    /// A user-defined trait type. The `Shared<TraitType>` field here refers to
+    /// the shared definition for each instance of this trait type. The `Vec<Type>`
+    /// represents the generic arguments (if any) to this trait type.
+    Trait(Shared<TraitType>, Vec<Type>),
+
```

Comment:
```suggestion
```
Traits are not types. `impl Trait` or `dyn Trait` would be types but we don't support those yet